netaddr_test.go

```diff
@@ -17,186 +17,268 @@ import (
+			ipaddr: &net.IPAddr{IP: net.ParseIP("0.0.0.0")},
+		},
+		// Basic non-zero IPv4 address.
+		{
+			in:     "192.168.140.255",
+			ip:     IP{0, 0xffffc0a88cff, z4},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("192.168.140.255")},
+		},
+		// IPv4 address in windows-style "print all the digits" form.
+		{
+			in:     "010.000.015.001",
+			ip:     IP{0, 0xffff0a000f01, z4},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("10.0.15.1")},
+			str:    "10.0.15.1",
+		},
+		// IPv4 address with a silly amount of leading zeros.
+		{
+			in:     "000001.00000002.00000003.000000004",
+			ip:     IP{0, 0xffff01020304, z4},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("1.2.3.4")},
+			str:    "1.2.3.4",
+		},
+		// Basic zero IPv6 address.
+		{
+			in:     "::",
+			ip:     IP{0, 0, z6noz},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("::")},
+		},
+		// Localhost IPv6.
+		{
+			in:     "::1",
+			ip:     IP{0, 1, z6noz},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("::1")},
+		},
+		// Fully expanded IPv6 address.
+		{
+			in:     "fd7a:115c:a1e0:ab12:4843:cd96:626b:430b",
+			ip:     IP{0xfd7a115ca1e0ab12, 0x4843cd96626b430b, z6noz},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("fd7a:115c:a1e0:ab12:4843:cd96:626b:430b")},
+		},
+		// IPv6 with elided fields in the middle.
+		{
+			in:     "fd7a:115c::626b:430b",
+			ip:     IP{0xfd7a115c00000000, 0x00000000626b430b, z6noz},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("fd7a:115c::626b:430b")},
+		},
+		// IPv6 with elided fields at the end.
+		{
+			in:     "fd7a:115c:a1e0:ab12:4843:cd96::",
+			ip:     IP{0xfd7a115ca1e0ab12, 0x4843cd9600000000, z6noz},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("fd7a:115c:a1e0:ab12:4843:cd96::")},
+		},
+		// IPv6 with the trailing 32 bits written as IPv4 dotted decimal.
+		{
+			in:     "::ffff:192.168.140.255",
+			ip:     IP{0, 0x0000ffffc0a88cff, z6noz},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("::ffff:192.168.140.255")},
+			str:    "::ffff:c0a8:8cff",
+		},
+		// IPv6 with a zone specifier.
+		{
+			in:     "fd7a:115c:a1e0:ab12:4843:cd96:626b:430b%eth0",
+			ip:     IP{0xfd7a115ca1e0ab12, 0x4843cd96626b430b, intern.Get("eth0")},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("fd7a:115c:a1e0:ab12:4843:cd96:626b:430b"), Zone: "eth0"},
+		},
+		// IPv6 with dotted decimal and zone specifier.
+		{
+			in:     "1:2::ffff:192.168.140.255%eth1",
+			ip:     IP{0x0001000200000000, 0x0000ffffc0a88cff, intern.Get("eth1")},
+			ipaddr: &net.IPAddr{IP: net.ParseIP("1:2::ffff:192.168.140.255"), Zone: "eth1"},
+			str:    "1:2::ffff:c0a8:8cff%eth1",
+		},
 	}
-	for _, s := range tests {
-		t.Run(s, func(t *testing.T) {
-			ip, err := ParseIP(s)
+
+	for _, test := range validIPs {
+		t.Run(test.in, func(t *testing.T) {
+			got, err := ParseIP(test.in)
 			if err != nil {
 				t.Fatal(err)
 			}
-			ip2, err := ParseIP(s)
+			if got != test.ip {
+				t.Errorf("ParseIP(%q) got %#v, want %#v", test.in, got, test.ip)
+			}
+
+			// Check that parsing is stable.
+			got2, err := ParseIP(test.in)
```

Comment:
Stable as in, there is no intermediate state established in the underlying parser that results in a different result when parsing the same input string again? I would have expected something like `got.String()` as an input to test round tripping.

Reply:
Yeah, I guess strictly speaking this is a test that ParseIP is a pure function.

Reply:
Added a test that `ParseIP(ip.String()) == ip`.